### PR TITLE
Suggestion: detection, spot fix, jvm-wide blacklist from config file, jvm-wide whitelist

### PR DESCRIPTION
**Pull request by tsohlacol**
In the modified form, Deserial Foil will work under more use cases: detection, spot-band-aid, spot-fix, updated JVM-wide blacklist, and JVM-wide whitelist.   For **detection**, the change would enable filterable logging of deserialization events, along with a script to parse the log results.  This enables companies to test if they’re affected by the vulnerability in question, instead of blindly applying a fix to their production application and hoping that the fix does not cause stability issues.  For **spot-band-aid**, the contribution would allow developers to opt-in to the blacklist via a code change.  This way, any instability impact is limited to the locations in the JVM where the fix was applied, not to all features in the JVM.   For **spot-fix**, the same code can be configured to properly fix the vulnerability by whitelisting only the deserialization events that they want to happen; again, impact of instability is limited to only those locations that opt-in.  For **JVM-wide blacklist**, the Akamai contribution moves the blacklisted items from hard coded in the source code to a configuration file, so that the only thing needed to blacklist a newly discovered attack vector is a configuration change and application restart, instead of a rebuild.  For **JVM-wide whitelist**, the suggested change creates a whitelist in parallel to the pre-existing JVM-wide blacklist, affecting the application administrator to apply a true fix across the entire application, if they choose to accept the stability risk.  They also can reconfigure the JVM-wide fix without a new build – they simply need to reconfigure with a configuration file.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -13,23 +13,58 @@ available for a long time. Because of the nature of the vulnerability, some
...
+// = new SafeObjectInputStream(inputStream, SafeObjectInputStream.BLACKLIST);
+in.addToList(ClassThatIsSafeToDeserialize.getName());
+in.addToList("com.my.SafeDeserializable");
+
+// then just use like normal
+in.readObject();
+```
+
+### Reporting on Serialization Usage
+The shim can be instructed to report when serialization occurs.  This allows you to determine where in your application deserialation is actually occuring - assuming that you exercise the relevant functionality.
```
Comment (nahsra) [inline README.md]: Spelling -- "deserialization".

File: README.md [hunk 2]
```diff
@@ -13,23 +13,58 @@ available for a long time. Because of the nature of the vulnerability, some
...
+
+To use this, built it just as described in "JVM-wide fix" but, use the following command line options instead:
+
 ```
 -javaagent:/path/to/contrast-rO0.jar
 ```
```
Comment (nahsra) [inline README.md:65]: > =( I like this line.

File: README.md [hunk 3]
```diff
@@ -39,8 +74,87 @@ for legitimate purposes during deserialization.
...
+                           events if @com.danga.Memcached is included in the 
+                           config file and ignoreStack is enabled., memcached 
+                           should be much less pronounced in the logs.  Likely 
+                           to have noticeable performance impact.  Can be 
+                           specified in combination with -DrO0.gnoreClasses.  
+                           Only useful if reporting is enabled.
+                           
+-DrO0.outfile=filename     Used to control where output from this utility is sent.
+                           If no output is specified, it will go to System.out.  If
+                           filename is specified, deserial foil will try to write
```
Comment (nahsra) [inline README.md]: We should change all references to "deserial foil" to "rO0".

File: README.md [hunk 4]
```diff
@@ -48,6 +162,6 @@ Although it's not tested on them all, the agent should work well on the followin
 * OpenJDK/HotSpot, JRockit, IBM
 
 ## Who made this?
-This project is sponsored by [Contrast Security](http://www.contrastsecurity.com/) and released under the BSD license for developers.
+This project is sponsored by [Contrast Security](http://www.contrastsecurity.com/) and released under the BSD license for developers.  It includes contributions from Akamai Technologies.
```
Comment (nahsra) [inline README.md:168]: I'm happy to include the Akamai logo if branding guidelines allow. Up to you.

File: config/unit_test_config_2 [hunk 5]
```diff
@@ -0,0 +1,12 @@
+############
+# HELP
+#
+# First chracter	Meaning
+# --------------	-------
+#	#		comment
+#	-		blacklist
+#	+		whitelist
+#	$		ignore during reporting if CLASS matcheso
```
Comment (nahsra) [inline config/unit_test_config_2]: matcheso -> matches (unless spoken with italian accent)

File: run_tests.sh [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+echo "recompiling"
+javac ../src/main/java/com/contrastsecurity/rO0/*.java ../src/main/java/com/contrastsecurity/rO0/TestCases/*.java ../src/main/java/com/akamai/security/*.java -classpath ../lib/asm-5.0.4.jar:../lib/asm-commons-5.0.4.jar  -d .
+
+echo "creating jar files"
+jar cfm contrast-test.jar ../mf/test.mf    com/contrastsecurity/rO0/TestCases/*.class
+jar cfm contrast-ro0.jar  ../mf/contrast.mf com/contrastsecurity/rO0/*.class
+jar cfm contrast-ro0-spotfix.jar ../mf/safeois.mf com/akamai/security/*.class
+
+echo "executing tests"
+java -javaagent:/Users/akatz/git/rO0/contrast-ro0/bin/contrast-ro0.jar -Xbootclasspath/p:"/Users/akatz/git/rO0/contrast-ro0/bin/contrast-rO0.jar:/Users/akatz/git/rO0/contrast-ro0/lib/asm-5.0.4.jar:/Users/akatz/git/rO0/contrast-ro0/lib/asm-commons-5.0.4.jar" -jar contrast-test.jar -Dfile.encoding=UTF-8 -classpath /Users/akatz/git/rO0/contrast-ro0/bin:/Users/akatz/git/rO0/contrast-ro0/lib/asm-5.0.4.jar:/Users/akatz/git/rO0/contrast-ro0/lib/asm-commons-5.0.4.jar
```
Comment (nahsra) [inline run_tests.sh]: Is this mean to be an integration test, runnable by others? If so, should we make the home directory dynamic?

File: src/main/java/com/contrastsecurity/rO0/DeserialFoilConfig.java [hunk 7]
```diff
@@ -0,0 +1,358 @@
+package com.contrastsecurity.rO0;
+
+import java.io.BufferedReader;
+import java.io.FileNotFoundException;
+import java.io.FileReader;
+import java.io.IOException;
+import java.util.Vector;
+
+public class DeserialFoilConfig {
```
Comment (nahsra) [inline src/main/java/com/contrastsecurity/rO0/DeserialFoilConfig.java]: Again, RO0Config?

File: src/main/java/com/contrastsecurity/rO0/DeserialFoilConfig.java [hunk 8]
```diff
@@ -0,0 +1,358 @@
...
+	}
+
+	@SuppressWarnings("rawtypes")
+	public boolean isBlacklisted(Class klass) {
+		return isBlacklisted(klass.getName());
+	}
+	
+	public boolean getWhitelistEnabled() { return this.whitelistEnabled; }
+	public boolean getBlacklistEnabled() { return this.blacklistEnabled; }
+	public boolean getClassIgnoreEnalbed() { return this.classIgnoreEnabled; }
```
Comment (nahsra) [inline src/main/java/com/contrastsecurity/rO0/DeserialFoilConfig.java]: Typo - getClassIgnoreEnabled().

File: src/main/java/com/contrastsecurity/rO0/RO0Agent.java [hunk 9]
```diff
@@ -20,22 +34,81 @@ public static void agentmain(String args, Instrumentation inst) throws IOExcepti
 	private static void setup(String args, Instrumentation inst) {
 		ClassFileTransformer xform = new RO0Transformer();
 		inst.addTransformer(xform);
+		readEnvironment();
+	}
+	
+	private static void readEnvironment() {
+		
+		String outfile = System.getProperty("RO0.outfile");
```
Comment (nahsra) [inline src/main/java/com/contrastsecurity/rO0/RO0Agent.java]: Would love for our System properties to be "RO0" -> "rO0". I only made it with a capital R when I had to -- when forced by Java's rules about class names.

File: README.md (hunk 2 above)
Comment (tsohlacol) [inline README.md:65]: ~~I'm not sure why I would have removed that...  will re-add.~~

The line actually still exists at line 34 of this edit (though slightly modified).  If you want the original text, I'm happy to revert the line :)

File: src/main/java/com/contrastsecurity/rO0/RO0Agent.java (hunk 9 above)
Comment (tsohlacol) [inline src/main/java/com/contrastsecurity/rO0/RO0Agent.java]: Yeah, that was a typo, and contradicts the README.  Will be fixed in my next submit.

File: README.md (hunk 1 above)
Comment (tsohlacol) [inline README.md]: should be fixed in next submit - along with several other spelling errors.

File: README.md (hunk 3 above)
Comment (tsohlacol) [inline README.md]: should be fixed in next submit.

File: README.md (hunk 4 above)
Comment (tsohlacol) [inline README.md:168]: Investigating.  Will let you know if this would be allowed.

File: config/unit_test_config_2 (hunk 5 above)
Comment (tsohlacol) [inline config/unit_test_config_2]: should be fixed in next submit.

File: run_tests.sh (hunk 6 above)
Comment (tsohlacol) [inline run_tests.sh]: should be fixed in next submit.

File: src/main/java/com/contrastsecurity/rO0/DeserialFoilConfig.java (hunk 7 above)
Comment (tsohlacol) [inline src/main/java/com/contrastsecurity/rO0/DeserialFoilConfig.java]: shoudl be fixed in next submit

File: src/main/java/com/contrastsecurity/rO0/DeserialFoilConfig.java (hunk 8 above)
Comment (tsohlacol) [inline src/main/java/com/contrastsecurity/rO0/DeserialFoilConfig.java]: should be fixed in next submit

**Pull request merged**